text/0000-service-hook.md

```diff
@@ -0,0 +1,104 @@
+* Start Date: 2019-03-21
+* RFC PR: (leave this empty)
+* Fusion Issue: (leave this empty)
+
+# Summary
+
+Add hook for consuming a registered service
+
+
+# Basic example
+
+We have a plugin registered to a token. We want to use the service that plugin provides.
+
+```javascript
+// components/Example.js
+import {ExampleToken} from 'fusion-tokens';
+import {useService} from 'fusion-react';
+
+export default function Example() {
+  const service = useService(ExampleToken);
+
+  return (
+    <button onClick={service}>Invoke service</button>
+  );
+}
+```
+
+# Motivation
+
+Consuming services through the legacy Context API is cumbersome and increasingly difficult with the adoption of hooks in Fusion. This proposal aims to provide a simplified and contemporary way of providing services to a React application.
+
+This work will not include refactoring existing plugins to hooks but will allow the migration path for all current '-react' plugins in the near future.
+
+# Detailed design
+
+The `useService` hook would get a function from Context to lookup the registered service for the supplied token.
+
+```javascript
+// fusion-react
+const ServiceContext = React.createContext();
+
+export function useService(token: Token<any>): any {
```

Comment:
I think we will want the hook to take a generic. Something like this: https://flow.org/try/#0KYDwDg9gTgLgBDAnmYcCiIZQIYGMYBKwMArlAHYAqyqAvHADwBqAfABRsCUctLcT3XvwDcAKFCRYCGnEoQA1sHINKfegG9RcOFwBcsgDRa4EMDACWEctgA2+rjz4A3COYAmcAD6HjAZxh48r76AIJQOIgMmtraAPSxcMC+NubkMAC0bua+2ABGNsDp5KAZKcXG2kgouhUx3gDkMApK9bXaDWA2JADmqa0xdXD1UMC9-sBQ-QNeQ0oAFtjkuMBTAw225ti+6QBmUBAAtquD9Rtb6U31RgP+gfr+UKnd13AAvixGr2IDojskSxYrHBcCNsDBgHJFMpVGxrAdgPcsE9OPpIUoVHxotpcFZ-AhmuR9ItEDw4OQSDYbN9tCNSBR8VCxK9aqIceQ8WjyCFSSDgGCIQSGA8nux6nNgJSIPVONTtOJwNB4H8AZZyHASL5gABlCZOczLFSowXEljsJpQ1EouAAEgAwrYbAwMFg8IRiGQqDQDLJMcZ4rIAPIAEQDcHMB06wHhaVEzOmcFZuPgmqgeuWpI12t1+uADCqwAgO1kBJC7E5IRlQA

We have some of those helpers already in `fusion-core`

Reply:
Agreed that we should be using generics here to ensure we can type the return type.  I'd also propose an alternative:

```
export function useService<TService>(token: Token<TService>): TService {
  // impl
}

// usage
const service = useService<string>(SomeStringToken); // typed as 'string'
```

The consumer doesn't care what the token is typed as, but rather the service provided.

Pros:
* User gets what they ask for.
* We can avoid the utility function magic of extracting the service type from the token.
* Avoids usage of `any` here.
* Explicit usage of types preferable over `typeof`.

Cons:
* Types for services need to be exported in order to be used.  We aren't great at this today.
* Can't provide a subtype as the generic type (a gripe I have with Flow :/)

Reply:
I definitely agree that it makes more sense from the consumer. They are interested in a return value, not an arbitrary and obligatory `typeof`.

I was trying to find situations where this would not cover what's actually happening. The only thing I can find at this point is if someone used `app.enhance` to completely change the returned service. It seems like this is somehow not disallowed. You have to properly forward the service through an `enhance` call.

To go off of the cons, there may be some service types that are exhausting to write by hand. Since these aren't being exported I could see that being a big downside. It's easier to use `typeof` though less explicit.